java/src/main/native/src/RmmJni.cpp

```diff
@@ -14,14 +14,118 @@
  * limitations under the License.
  */
 
+#include <rmm/mr/device/default_memory_resource.hpp>
+
 #include "jni_utils.hpp"
 
+namespace {
+
+class java_event_handler_memory_resource : public rmm::mr::device_memory_resource {
+public:
+  java_event_handler_memory_resource(JNIEnv* env, jobject jhandler) {
+    if (env->GetJavaVM(&jvm) < 0) {
+      throw cudf::jni::jni_exception("GetJavaVM failed");
+    }
+    jclass cls = env->GetObjectClass(jhandler);
+    if (cls == nullptr) {
+      throw cudf::jni::jni_exception("class not found");
+    }
+    on_alloc_fail_method = env->GetMethodID(cls, "onAllocFailure", "(J)Z");
+    if (on_alloc_fail_method == nullptr) {
+      throw cudf::jni::jni_exception("onAllocFailure method");
+    }
+    handler_obj = env->NewGlobalRef(jhandler);
+    if (handler_obj == nullptr) {
+      throw cudf::jni::jni_exception("global ref");
+    }
+
+    // The previous resource could be null, but getting the current resource never is.
+    resource = rmm::mr::get_default_resource();
+    rmm::mr::device_memory_resource* old = rmm::mr::set_default_resource(this);
```

Comment:
I've been referring to these kinds of `device_memory_resource` implementations that take and existing resource and layer some functionality on to it as "resource adaptors". E.g., see the [`logging_resource_adaptor`](https://github.com/rapidsai/rmm/blob/f161ca90e45a2f0982fa7222dc99dd8dda859bc0/include/rmm/mr/device/logging_resource_adaptor.hpp)

In adaptors, instead of just grabbing the default resource in the ctor, I'd recommend making the "upstream" resource a parameter of the ctor (see the logging adaptor). That makes the side-effects more obvious via dependency injection. Otherwise it could be surprising that constructing a `java_event_handle_memory_resource` is changing the default resource. 

So the callsite to `java_event_handler_memory_resource` would look like:
```
auto old = rmm::mr:get_default_resource();
auto java = java_event_handler_memory_resource(..., old);
rmm::mr::set_default_resource(&java);
```